backend/core/rule/tree.py

```diff
@@ -344,23 +344,36 @@ def tree_into_dict(self, hide=False):
     def add_children(self, obj):
         self.children.append(obj)
 
-    def eval_children(self, taken_list):
+    def eval_children(self, taken_list, codeset=None):
         """Evaluate children recursively
 
         Evaluate TreeNode and set self.data to its value.
         If children exist, evaluate them all recursively.
 
         Args:
             taken_list: taken course list
+            codeset: dictionary of list of course code containing
+                    code itself and replacable codes.
+                    Do not set this argument and let it be default.
 
         Returns:
             None
         """
 
+        # If codeset is empty, construct it first
+        if codeset is None:
+            codeset = {}
+            # iterate over taken list, and lookup Replace model
+            for course in taken_list:
+                code = course['code']
+                # place its code in head of list, so that the replace code
+                # acts as a fallback of original course code
+                codeset[code] = [code] + [replace.to_code for replace in Replace.objects.filter(from_code=code)]
```

Comment:
And, filtering module with list argument is not valid. Because Django generates SQL queries for the `filter` method, so I don't think that such searching method is exists.